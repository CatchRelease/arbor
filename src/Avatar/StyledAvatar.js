import React from 'react';
import { withTheme } from 'emotion-theming';
import PropTypes from 'prop-types';
import Flex from '../Flex';

import { sizes } from '../theme/avatars';

export const StyledAvatar = ({
  baseColor,
  border,
  size,
  subtle,
  theme: { avatarSizes }, // eslint-disable-line react/prop-types
  ...props
}) => {
  const bg = subtle ? `monochrome.white` : baseColor;
  const borderColor = baseColor;

  return (
    <Flex
      {...{
        width: avatarSizes[size],
        height: avatarSizes[size],
        bg,
        border,
        borderColor,
        ...props
      }}
    />
  );
};

StyledAvatar.propTypes = {
  baseColor: PropTypes.string.isRequired,
  border: PropTypes.string.isRequired,
  size: PropTypes.oneOf(Object.keys(sizes)).isRequired,
  subtle: PropTypes.bool.isRequired
};

export default withTheme(StyledAvatar);
