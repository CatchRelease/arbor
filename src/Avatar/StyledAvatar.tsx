import { forwardRef } from 'react';

import { Theme, withTheme } from '@emotion/react';

import Flex from '../Flex';
import { Size } from '../theme/avatars';

type Props = {
  baseColor: string;
  border: string;
  size: Size;
  subtle: boolean;
  theme: Theme;
};

export const StyledAvatar = forwardRef<HTMLDivElement, Props>(
  (
    { baseColor, border, size, subtle, theme: { avatarSizes }, ...props },
    ref
  ) => {
    const bg = subtle ? `monochrome.white` : baseColor;
    const borderColor = baseColor;

    return (
      <Flex
        {...{
          width: avatarSizes[size],
          height: avatarSizes[size],
          bg,
          border,
          borderColor,
          ref,
          ...props
        }}
      />
    );
  }
);

export default withTheme(StyledAvatar);
