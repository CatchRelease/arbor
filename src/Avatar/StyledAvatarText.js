import React from 'react';
import PropTypes from 'prop-types';
import { withTheme } from 'emotion-theming';

import Text from '../Text';
import { sizes } from '../theme/avatars';

export const StyledAvatarText = ({
  size,
  subtle,
  theme: { avatarFontSizes },
  baseColor,
  ...props
}) => {
  const color = subtle ? `${baseColor}.darker` : 'monochrome.white';
  const fontSize = avatarFontSizes[size];

  return <Text {...{ color, fontSize, ...props }} />;
};

StyledAvatarText.propTypes = {
  size: PropTypes.oneOf(Object.keys(sizes)).isRequired
};

export default withTheme(StyledAvatarText);
