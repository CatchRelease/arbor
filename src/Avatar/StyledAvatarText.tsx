import { FC } from 'react';

import { Theme, withTheme } from '@emotion/react';

import Text from '../Text';
import { Size } from '../theme/avatars';

type Props = {
  baseColor: string;
  size: Size;
  subtle: boolean;
  theme: Theme;
};

export const StyledAvatarText: FC<Props> = ({
  baseColor,
  size,
  subtle,
  theme: { avatarFontSizes },
  ...props
}) => {
  const color = subtle ? baseColor : 'monochrome.white';
  const fontSize = avatarFontSizes[size];

  return <Text {...{ color, fontSize, ...props }} />;
};

export default withTheme(StyledAvatarText);
