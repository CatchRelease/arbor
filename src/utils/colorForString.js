const colorForString = (string, colors) => {
  const code = string
    .split('')
    .reduce((sum, char) => sum + char.charCodeAt(0), 0);
  return colors[code % colors.length];
};

export default colorForString;
