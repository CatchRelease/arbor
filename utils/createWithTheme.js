import React from 'react';
import { CacheProvider } from '@emotion/core';
import createCache from '@emotion/cache';
import { create } from 'react-test-renderer';
import { ThemeProvider } from 'emotion-theming';
import theme from '../src/theme';

const myCache = createCache();
myCache.compat = true;

export default (Component) =>
  create(
    <CacheProvider value={myCache}>
      <ThemeProvider theme={theme}>{Component}</ThemeProvider>
    </CacheProvider>
  );
