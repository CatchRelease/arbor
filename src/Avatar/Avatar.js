import React from 'react';
import PropTypes from 'prop-types';

import { colorForString } from '../utils';
import StyledAvatar from './StyledAvatar';
import StyledAvatarText from './StyledAvatarText';

const AVATAR_COLORS = [
  'palette.blue',
  'palette.red',
  'palette.yellow',
  'palette.green'
];

const getInitials = name => {
  const [firstName, ...additionalNames] = name.split(' ');
  const firstInitial = firstName[0];

  if (additionalNames.length) {
    const lastName = additionalNames[additionalNames.length - 1];
    const lastInitial = lastName[0];
    return `${firstInitial}${lastInitial}`;
  }

  return firstInitial;
};

const Avatar = ({ name, subtle, ...props }) => {
  const baseColor = colorForString(name, AVATAR_COLORS);

  return (
    <StyledAvatar
      {...{
        alignItems: 'center',
        baseColor,
        border: '1px solid',
        borderRadius: '50%',
        justifyContent: 'center',
        subtle,
        ...props
      }}
    >
      <StyledAvatarText {...{ ...props, baseColor, subtle }}>
        {getInitials(name).toUpperCase()}
      </StyledAvatarText>
    </StyledAvatar>
  );
};

Avatar.propTypes = {
  /**
   * The name for the user's avatar. The name is used to determine the color for
   * the avatar component. See utils/colorForString.
   *
   * The initials for display will also be determined based of the name
   * provided.
   */
  name: PropTypes.string.isRequired,

  /**
   * Use a subtle version of the avatar's color styling.
   * */
  subtle: PropTypes.bool
};

Avatar.defaultProps = {
  subtle: false
};

export default Avatar;
