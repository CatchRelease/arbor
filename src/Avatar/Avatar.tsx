import { forwardRef } from 'react';

import { Size } from '../theme/avatars';
import { palette } from '../theme/colors';
import { colorForString } from '../utils';
import StyledAvatar from './StyledAvatar';
import StyledAvatarText from './StyledAvatarText';

const AVATAR_COLORS = Object.keys(palette).map(
  (color) => `palette.${color}.default`
);

const getInitials = (name: string) => {
  const [firstName, ...additionalNames] = name.split(' ');
  const firstInitial = firstName[0];

  if (additionalNames.length) {
    const lastName = additionalNames[additionalNames.length - 1];
    const lastInitial = lastName[0];
    return `${firstInitial}${lastInitial}`;
  }

  return firstInitial;
};

const getText = (name: string) =>
  name.match(/[A-Za-z]/) ? getInitials(name).toUpperCase() : name;

type Props = {
  baseColor?: string;
  name: string;
  size?: Size;
  subtle?: boolean;
};

const Avatar = forwardRef<HTMLDivElement, Props>(
  (
    {
      name,
      subtle = false,
      size = 'default',
      baseColor: baseColorProp,
      ...props
    },
    ref
  ) => {
    const baseColor = baseColorProp || colorForString(name, AVATAR_COLORS);

    return (
      <StyledAvatar
        {...{
          alignItems: 'center',
          baseColor,
          border: '1px solid',
          borderRadius: '50%',
          justifyContent: 'center',
          size,
          subtle,
          ref,
          ...props
        }}
      >
        <StyledAvatarText {...{ baseColor, size, subtle }}>
          {getText(name)}
        </StyledAvatarText>
      </StyledAvatar>
    );
  }
);

export default Avatar;
