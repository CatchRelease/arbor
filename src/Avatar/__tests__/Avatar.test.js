import React from 'react';
import { shallow } from 'enzyme';

import Avatar from '../Avatar';
import colorForString from '../../utils/colorForString';
import createWithTheme from '../../../utils/createWithTheme';

jest.mock('../../utils/colorForString.js');

describe('<Avatar />', () => {
  it('renders properly', () => {
    const tree = createWithTheme(<Avatar name="Bruce Wayne" />);

    expect(tree).toMatchSnapshot();
  });

  describe('initials', () => {
    it('renders capital letters for initials', () => {
      const avatar = shallow(<Avatar name="bruce wayne" />);

      expect(avatar).toHaveText('BW');
    });

    context('single word provided for name', () => {
      it('renders the first letter of the word', () => {
        const avatar = shallow(<Avatar name="Batman" />);

        expect(avatar).toHaveText('B');
      });
    });

    context('multiple words provided for name', () => {
      it('renders the first letter of each word', () => {
        const avatar = shallow(
          <Avatar name="Alfred Thaddeus Crane Pennyworth" />
        );

        expect(avatar).toHaveText('AP');
      });
    });
  });

  describe('background color', () => {
    it('is calculated by the name property', () => {
      colorForString.mockReturnValue('blue');
      const avatar = shallow(<Avatar name="Batman" />);

      expect(avatar.find('StyledAvatar')).toHaveProp({
        bg: 'blue'
      });
    });
  });
});
