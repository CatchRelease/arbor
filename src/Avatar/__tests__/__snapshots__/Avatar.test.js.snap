// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> renders properly 1`] = `
.emotion-2 {
  box-sizing: border-box;
  border-color: #29863A;
  border-radius: 50%;
  border: 1px solid;
  background-color: #29863A;
  height: 24px;
  width: 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  font-family: Motiva Sans;
  color: #FFFFFF;
  font-size: 0.75rem;
}

<div
  className="emotion-2 emotion-3"
  height="24px"
  width="24px"
>
  <p
    className="emotion-0 emotion-1"
    color="monochrome.white"
    fontSize="0.75rem"
  >
    BW
  </p>
</div>
`;
